src/lightgbm/src/main/scala/TrainUtils.scala

```diff
@@ -30,102 +30,167 @@ private object TrainUtils extends Serializable {
+      val weights = rows.map(row => row.getDouble(row.fieldIndex(col)))
+      addField(weights, "weight", numRows, datasetPtr)
+    }
+    datasetPtr
+  }
 
-      if (weightColumn.isDefined) {
-        val weights = rows.map(row => row.getDouble(row.fieldIndex(weightColumn.get)))
-        addField(weights, "weight", numRows, datasetPtr)
-      }
+  def createBooster(trainParams: TrainParams, trainDatasetPtr: Option[SWIGTYPE_p_void],
+                    validDatasetPtr: Option[SWIGTYPE_p_void]): Option[SWIGTYPE_p_void] = {
+    // Create the booster
+    val boosterOutPtr = lightgbmlib.voidpp_handle()
+    val parameters = trainParams.toString()
+    LightGBMUtils.validate(lightgbmlib.LGBM_BoosterCreate(trainDatasetPtr.get, parameters, boosterOutPtr), "Booster")
+    val boosterPtr = Some(lightgbmlib.voidpp_value(boosterOutPtr))
+    trainParams.modelString.foreach { modelStr =>
+      val booster = LightGBMUtils.getBoosterPtrFromModelString(modelStr)
+      LightGBMUtils.validate(lightgbmlib.LGBM_BoosterMerge(boosterPtr.get, booster), "Booster Merge")
+    }
+    validDatasetPtr.foreach { dataset =>
+      LightGBMUtils.validate(lightgbmlib.LGBM_BoosterAddValidData(boosterPtr.get,
+        dataset), "Add Validation Dataset")
+    }
+    boosterPtr
+  }
 
-      var boosterPtr: Option[SWIGTYPE_p_void] = None
+  def saveBoosterToString(boosterPtr: Option[SWIGTYPE_p_void], log: Logger): String = {
+    val bufferLength = LightGBMConstants.defaultBufferLength
+    val bufferLengthPtr = lightgbmlib.new_longp()
+    lightgbmlib.longp_assign(bufferLengthPtr, bufferLength)
+    val bufferLengthPtrInt64 = lightgbmlib.long_to_int64_t_ptr(bufferLengthPtr)
+    val bufferOutLengthPtr = lightgbmlib.new_int64_tp()
+    lightgbmlib.LGBM_BoosterSaveModelToStringSWIG(boosterPtr.get, 0, -1, bufferLengthPtrInt64, bufferOutLengthPtr)
+  }
+
+  case class BestEvaluation(evalCounts: Int, evalNames: Array[String], bestScore: Array[Double],
+                            bestScores: Array[Array[Double]], bestIter: Array[Int])
+
+  def createBestEval(boosterPtr: Option[SWIGTYPE_p_void]): Option[BestEvaluation] = {
+    // Need to keep track of best scores for each metric, see callback.py in lightgbm for reference
+    val evalCountsPtr = lightgbmlib.new_intp()
+    val resultCounts = lightgbmlib.LGBM_BoosterGetEvalCounts(boosterPtr.get, evalCountsPtr)
+    LightGBMUtils.validate(resultCounts, "Booster Get Eval Counts")
+    val evalCounts = lightgbmlib.intp_value(evalCountsPtr)
+    val bestScore = new Array[Double](evalCounts)
+    val bestScores = new Array[Array[Double]](evalCounts)
+    val bestIter = new Array[Int](evalCounts)
+    // For debugging, can get metric names:
+    val evalNamesPtr = lightgbmlib.LGBM_BoosterGetEvalNamesSWIG(boosterPtr.get, evalCounts)
+    val evalNames = (0 until evalCounts).map(lightgbmlib.stringArray_getitem(evalNamesPtr, _)).toArray
+    Some(BestEvaluation(evalCounts, evalNames, bestScore, bestScores, bestIter))
```

Comment:
We should chat in person. You really just need eval names here

Reply:
sure, feel free to refactor this more :)